jetbrains-rider/src/software/aws/toolkits/jetbrains/services/lambda/dotnet/DotNetSamDebugSupport.kt

```diff
@@ -0,0 +1,305 @@
+import software.aws.toolkits.jetbrains.services.lambda.execution.local.SamDebugSupport
+import software.aws.toolkits.jetbrains.services.lambda.execution.local.SamRunningState
+import software.aws.toolkits.resources.message
+import java.io.File
+import java.io.OutputStream
+import java.util.Timer
+import kotlin.concurrent.schedule
+
+/**
+ * Debugger is launched via SAM CLI command: '$ sam local invoke' and pass debugger parameters --debugger-path and --debugger-args.
+ * Under SAM CLI this command is translated into:
+ * - volume that is mounted in Docker to launch debugger (--debugger-path)
+ * - additional debugger arguments that are applied when launching debugger inside Docker (--debugger-args):
+ *    '$ dotnet <debug_args_list> MockBootstraps.dll --debugger-spin-wait'
+ *
+ * Rider uses it's own Debugger Worker process that run under .NET with extra parameters to configure:
+ * - path to runtime
+ * - path to DLL
+ * - debugger mode (client/server)
+ * - frontend port
+ * - backend port
+ *
+ * To launch Rider Debugger under Docker we have created a simple dotnet core app that launch Debugger Worker instance
+ * using Mono runtime that is taken from Rider SDK (runtime.sh).
+ * This app is hosted in github repo - https://github.com/JetBrains/JetBrains.Rider.Debugger.Launcher and is
+ * delivered via GutHub releases with packed output.
+ * This output is copied to the same Rider SDK path that is mounted in Docker (app should be placed under "Amazon.Runner" folder).
+ *
+ * After Debugger Worker process is started on Docker instance it initializes ports and process to run - MockBootstraps.dll
+ * (the internal compiled with lambda code)
+ */
+class DotNetSamDebugSupport : SamDebugSupport {
+    companion object {
+        private val logger = getLogger<DotNetSamDebugSupport>()
+
+        private const val debuggerDir = "/tmp/lambci_debug_files"
+        private const val dotnetCorCliPath = "/var/lang/bin/dotnet"
+        private const val mockBootstrapsPath = "/var/runtime/MockBootstraps.dll"
+    }
+
+    /**
+     * Check whether the Rider Debugger Worker Launcher app (that is required to run Debugger) is downloaded into Rider SDK.
+     */
+    override fun isSupported(): Boolean {
+        val debuggerWorkerAssembly = RiderEnvironment.getBundledFile(DebuggerWorkerPlatform.AnyCpu.assemblyName)
+        val binDir = debuggerWorkerAssembly.parentFile
+        val debugLauncherFile = File(binDir, "Amazon.Runner").resolve("JetBrains.Debugger.Launcher.dll")
+
+        val debuggerLauncherExists = debugLauncherFile.exists()
+        if (!debuggerLauncherExists) {
+            logger.error { "JetBrains.Rider.Debugger.Launcher runnable does not exists" }
+        }
+        return debuggerLauncherExists
+    }
+
+    override fun patchCommandLine(debugPort: Int, state: SamRunningState, commandLine: GeneralCommandLine) {
+        val debuggerWorkerAssembly = RiderEnvironment.getBundledFile(DebuggerWorkerPlatform.AnyCpu.assemblyName)
```

Comment:
Is this self contained? Or does mono keep to be in the container?

Reply:
We do not need mono inside container. We launch DebuggerWorker process under mono that we bundle in Rider SDK. We get access to Mono through Docker volume directory that we bind to Rider SDK directory with DebuggerWorker.

Reply:
Does this work cross OS? What about if parent OS is Windows?

Reply:
Here, we rely on fact that SAM CLI starts Linux instance. If it might change we should rather update the JetBrains.Rider.Debugger.Launcher app to detect an instance OS and run corresponding DebuggerWorker executable.

Reply:
Ah. No it's Linux. Was just more concerned if DebuggerWorker is always compiled for Linux, even if Rider is running on Windows.